File: 0207-course-schedule/0207-course-schedule.cpp
typedef vector<vector<int>> GRAPH;

class Solution
{
public:
    void addDirectedEdge(GRAPH &adjList, int from, int to)
    {
        adjList[from].push_back(to);
    }

    bool topSort(GRAPH &adjList)
    {
        vector<int> inDeg(adjList.size(), 0);
        queue<int> q;
        vector<int> order;
        for (auto i : adjList)
        {
            for (int neighbour : i)
            {
                inDeg[neighbour]++;
            }
        }

        for (int i = 0; i < inDeg.size(); i++)
        {
            if (!inDeg[i])
                q.push(i);
        }

        while (!q.empty())
        {
            int node = q.front();
            q.pop();
            order.push_back(node);
            for (auto neihbour : adjList[node])
            {
                if (--inDeg[neihbour] == 0)
                    q.push(neihbour);
            }
        }

        if (order.size() != adjList.size())
            return false;
        return true;
    }
    bool canFinish(int numCourses, vector<vector<int>> &prerequisites)
    {
        GRAPH adjList(numCourses);
        for (auto i : prerequisites)
        {
            addDirectedEdge(adjList, i[1], i[0]);
        }

        return topSort(adjList);
    }
};

File: 0207-course-schedule/0207-course-schedule.java
import java.util.*;

class Solution {
    HashMap<Integer, List<Integer>> graph = new HashMap<>();

    public boolean canFinish(int numCourses, int[][] prerequisites) {
        buildGraph(prerequisites);
        return topologicalSort(numCourses);
    }

    private boolean topologicalSort(int numCourses) {
        int[] inDeg = new int[numCourses];
        for (List<Integer> edges : graph.values()) {
            for (int node : edges) {
                ++inDeg[node];
            }
        }

        Queue<Integer> queue = new LinkedList<>();
        for (int i = 0; i < numCourses; i++) {
            if (inDeg[i] == 0)
                queue.add(i);
        }

        ArrayList<Integer> resultOrder = new ArrayList<>();
        while (!queue.isEmpty()) {
            int currentNode = queue.poll();
            resultOrder.add(currentNode);
            for (int neighbour :  graph.getOrDefault(currentNode, Collections.emptyList())) {
                if (--inDeg[neighbour] == 0) {
                    queue.add(neighbour);
                }
            }
        }

        return resultOrder.size() == numCourses;
    }


    void buildGraph(int[][] prerequisites) {
        for (int[] edge : prerequisites) {
            addDirectedEdge(edge);
        }
    }

    void addDirectedEdge(int[] edge) {
        graph.computeIfAbsent(edge[0], k -> new ArrayList<>()).add(edge[1]);
    }
}